New hover dropdown menu for media element

joannag6 opened the pull request:
## Summary
Added a new dropdown menu that appears when you click on the More icon (which appears on hover)

![Screen Shot 2020-05-04 at 12 35 33 PM](https://user-images.githubusercontent.com/15134432/80933144-c2a81f00-8e05-11ea-8f20-2e5a9cc15209.png)
![Screen Shot 2020-05-04 at 12 35 50 PM](https://user-images.githubusercontent.com/15134432/80933147-c5a30f80-8e05-11ea-8f64-e16bee612a49.png)
![Screen Shot 2020-05-04 at 1 11 45 PM](https://user-images.githubusercontent.com/15134432/80933886-e7ea5c80-8e08-11ea-81cc-393f270d301e.png)

## To-do
Currently, the menu has the "Edit" and "Delete" options which don't do anything yet. These are being tracked in #354 and #1319 respectively. This PR is just to add the Menu UI (which is set to not display by default - simple way to feature guard this since #1464 is not yet implemented).

winterDroid commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,111 @@
+  width: 28px;
+`;
+
+const Div = styled.div`
```
`Div` feels like a very unspecific name. Is there something more descriptive?

joannag6 replied:
Renamed to IconContainer

joannag6 committed:
Added dropdown menu for media elements

Renamed Div to IconContainer

Updated value to be lowercase

Fixed menuCallback

Never display hover dropdown (until PRs to implement Edit and Delete are merged)

Included #1464 in the comment to add more context.

Added storybook test and other changes

Updated stories to use imported images

Use useEffect instead to handle pointer events and dropdown events

Cleaned up code

use useCallback

use useCallback in dropdown menu as well

Don't show More Icon / Dropdown Menu by default

Reduced size of test image / video

renamed storybook file

Using feature flags

Fixed margin

spacedmonkey commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,111 @@
+    setShowMenu(false);
+    switch (value) {
+      case 'Edit':
+        // TODO(#354): Edit Media Metadata via Media Library Hover Menu
```
There is an existing api call for this that can found [here](https://github.com/google/web-stories-wp/blob/master/assets/src/edit-story/app/api/apiProvider.js#L179-L195) called `updateMedia`.

joannag6 replied:
Thanks! Will add this in a future PR.

spacedmonkey commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,111 @@
+        // TODO(#354): Edit Media Metadata via Media Library Hover Menu
+        break;
+      case 'Delete':
+        // TODO(#1319): Media Library - Delete via Dropdown Menu from Hover
```
A new REST API call will need to be written here.

joannag6 replied:
Delete API call is added in #1654 and will be called in a future PR :)

swissspidy commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,111 @@
+ */
+function DropDownMenu({ showDisplayIcon, menuCallback }) {
+  const options = [
+    { name: __('Edit', 'web-stories'), value: 'Edit' },
```
Nit: I'd make `value` lowercase

swissspidy commented on assets/src/edit-story/components/library/panes/media/mediaElement.js:
```diff
@@ -173,7 +176,24 @@ const MediaElement = ({
+    // TODO: Set to `true` when Edit and Delete have been fully implemented and merged.
+    // Currently we don't have feature flags (#1464) so this is a simple way to keep this feature
+    // hidden until fully implemented.
+    const onPointerEnterImage = () => setShowMoreIcon(false);
```
Instead of having `onPointerEnterImage` and `onPointerEnterVideo`, wouldn't it be easier to have 1 `pointerEnter`?

E.g.

```js
const onPointerEnter = useCallback(() => {
  // TODO: Set to `true` when Edit and Delete have been fully implemented and merged.
  // Currently we don't have feature flags (#1464) so this is a simple way to keep this feature
  // hidden until fully implemented.
  setShowMoreIcon(true);

  if (type === 'video') {
    setShowVideoDetail(false);
    if (mediaElement.current) {
      mediaElement.current.play();
    }
  }
}, [type]);

const onPointerLeave = useCallback(() => {
  setShowMoreIcon(false);

  if (type === 'video') {
    setShowVideoDetail(true);
    if (mediaElement.current) {
      mediaElement.current.pause();
      mediaElement.current.currentTime = 0;
    }
  }
}, [type]);
```

swissspidy replied:
btw, the whole `currentTime = 0` thing seems super confusing. We might want to remove that because of the jumps/flickers this causes.

joannag6 replied:
Changed this part a bit to use `useEffect` instead, WDYT? 

And agreed, removed `currentTime = 0` for `pause` - but kept for `stop`.

swissspidy commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,117 @@
+ * Get a More icon that displays a dropdown menu on click.
+ *
+ * @param {Object} props Component props.
+ * @param {boolean} props.showDisplayIcon If the More icon should be displayed
```
Maybe rename to `collapsed` to make less confusing?

joannag6 replied:
Hmm I'm not sure if `collapsed` is the right word, what do you think of `hidden`?

swissspidy replied:
If it can be hidden like it can right now,  then perhaps it shouldn't even be rendered in the first place?

joannag6 replied:
Renamed to `pointerEntered`.

swissspidy commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,117 @@
+  ];
+
+  const [showMenu, setShowMenu] = useState(false);
+  const moreRef = useRef();
```
`useState(!showDisplayIcon)`? (or `useState(!collapsed)`, as per above comment)

joannag6 replied:
This should always be initialised as `false`, as the menu does not show by default (until the icon is clicked on).

swissspidy replied:
Makes me think whether a storybook story for this area is useful. I feel like it could be useful for testing.

joannag6 replied:
Added one! (Also added a bunch of test assets, not sure if this is the best place for them to live.)

swissspidy commented on assets/src/edit-story/components/library/panes/media/mediaElement.js:
```diff
@@ -173,7 +176,24 @@ const MediaElement = ({
   const onClick = () => onInsert(resource, width, height);
 
+  /// Callback for when the More dropdown menu has been opened / closed.
+  const menuCallback = (isMenuOpen) => {
```
Use `useCallback`?

joannag6 replied:
Done

swissspidy commented on assets/src/edit-story/components/library/panes/media/mediaElement.js:
```diff
@@ -173,7 +176,24 @@ const MediaElement = ({
+      mediaElement.current.pause();
+      mediaElement.current.currentTime = 0;
+    } else if (!isMenuOpen) {
+      setShowMoreIcon(false);
```
This causes odd behavior:

1. Click on icon
1. Menu opens
1. Video stops playing
1. Click on icon again
1. Menu closes
1. Icon disappears
1. Video still not playing

What I'd expect:

1. Click on icon
1. Menu opens
1. Video stops playing
1. Click on icon again
1. Menu closes
1. Video continues playing

It gets weirder:

1. Click on icon
1. Menu opens
1. Video stops playing
1. Move cursor away from video
1. `currentTime` is reset
1. Move cursor over video again
1. Video continues playing, with menu open

I'd say the video should never continue playing as long as the menu is open.

joannag6 replied:
Redid this part to use `useEffect` instead.

swissspidy commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,117 @@
+import Popup from '../../../../components/popup';
+import { ReactComponent as More } from '../../../../icons/more_horiz.svg';
+
+const MoreIcon = styled(More)`
```
This needs to be a proper `button` for accessbility reasons.

I recommend defining it in `assets/src/edit-story/components/button/index.js` together with other buttons:

```js
import { ReactComponent as MoreIcon } from '../../../../icons/more_horiz.svg';

// ...

export const More = (props) => (
  <StyledButton {...props}>
    <MoreIcon />
  </StyledButton>
);
```

joannag6 replied:
Done.

swissspidy commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,117 @@
+  // Keep icon and menu displayed if menu is open (even if user's mouse leaves the area).
+  return (
+    // TODO: Remove `false &&` when Edit and Delete have been fully implemented and merged.
+    // Currently we don't have feature flags (#1464) so this is a simple way to keep this feature
```
Now we do! #1700 just got merged.

See `docs/feature-flags.md` for details.

Might wanna use the flag within `MediaElement` instead of here, not sure.

joannag6 replied:
Yay! Done 😊

joannag6 committed:
Add default flag provider to testutils for library components

joannag6 committed:
Hide more icon if media item hasnt been uploaded yet

joannag6 committed:
Hide dropdown menu on scroll

swissspidy commented on assets/src/edit-story/components/form/dropDown/list.js:
```diff
@@ -99,6 +99,12 @@ function DropDownList({
   );
 
+  useEffect(() => {
+    // Hide dropdown menu on scroll.
```
_scroll_ or _wheel_?

Per [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/wheel_event):

>  Don't confuse the `wheel` event with the `scroll` event. The default action of a `wheel` event is implementation-specific, and doesn't necessarily dispatch a `scroll` event.

Also, isn't that event fired dozens of times? Should the callback be throttled?

spacedmonkey replied:
Does this only happen on wheel? Feel like it would happen on scroll as well.

spacedmonkey replied:
I just confirmed, that wheel can not be used. The page up and page down buttons also cause the issue you are trying to fix here.

joannag6 replied:
Using `scroll` instead.

Set `options` to `true` so "the callback will only be invoked once after which the event listener will be removed".

joannag6 committed:
Use scroll instead of wheel event, also set options=true

joannag6 committed:
Set initial value of dropdown list to first option, this is so the correct value is focussed

joannag6 committed:
Hide More icon on menu close

joannag6 committed:
Separate out menuCallback functions

swissspidy commented on includes/Story_Post_Type.php:
```diff
@@ -397,7 +397,15 @@ public static function get_editor_settings() {
+				 * Issue: #1319 and #354
+				 * Creation date: 2020-05-20
+				 */
+				'dropdownMenu' => false,
```
Nit: `dropdownMenu` is very generic. Maybe `mediaDropdownMenu`?

swissspidy reviewed (approved):
LGTM with one nit.

swissspidy commented on assets/src/edit-story/components/library/panes/media/dropDownMenu.js:
```diff
@@ -0,0 +1,125 @@
+          ref={moreButtonRef}
+          width="28"
+          height="28"
+          onClick={onMenuOpen}
```
One thing: If I click on the "more" button to open the menu - should I be able to close it by clicking on it again?

joannag6 replied:
Nope, it should just stay open :)

joannag6 committed:
Rename dropdownMenu flag to mediaDropdownMenu
